### PR TITLE
docs: Add a plugin overview page to the contributors guide

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

N/A

## Rationale for this change

Make it easier for new contributors to understand how Comet works and where to start looking at code when adding new operators or expressions.

## What changes are included in this PR?

New "Comet Plugin Overview Page".

## How are these changes tested?

Manually tested publishing locally.

### CODE REVIEW

File: docs/source/contributor-guide/plugin_overview.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+When the V2 data source API is being used, `BatchScanExec` is replaced with `CometBatchScanExec`.
+
+## CometExecRule
+
+`CometExecRule` attempts to transform a Spark physical plan into a Comet plan.
+
+This rule traverses bottom-up from the original Spark plan and attempts to replace each node with a Comet equivalent. For example, a `ProjectExec` will be replaced by `CometProjectExec`.
+
+When replacing a node, various checks are performed to determine if Comet can support the operator and its expressions. If an operator or expression is not supported by Comet then the reason will be stored in a tag on the underlying Spark node. Running `explain` on a query will show any reasons that prevented the plan from being executed natively in Comet. If any part of the plan is not supported in Comet then the original Spark plan will be returned.
```
Comment (viirya) [inline docs/source/contributor-guide/plugin_overview.md:44]: I've not tried, but don't we need do anything to get the unsupported info from `explain`?

cc @parthchandra

**Review by viirya — APPROVED** [review]
One question about `explain` info. Otherwise looks good to me.

**Review by kazuyukitanimura — APPROVED** [review]

File: docs/source/contributor-guide/plugin_overview.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+Comet does not support partially replacing subsets of the plan because this would involve adding transitions to convert between row-based and columnar data between Spark operators and Comet operators and the overhead of this could outweigh the benefits of running parts of the plan natively in Comet.
```
Comment (sunchao) [inline docs/source/contributor-guide/plugin_overview.md:46]: nit: maybe we should mention this is within a Spark stage? instead of the whole Spark plan

**Review by sunchao — APPROVED** [review]
LGTM too with one nit

File: docs/source/contributor-guide/plugin_overview.md (hunk 1 above)
Comment (andygrove) [reply docs/source/contributor-guide/plugin_overview.md]: I removed the reference to `explain` here. I will add something in the future when I understand this part more.

File: docs/source/contributor-guide/plugin_overview.md (hunk 2 above)
Comment (andygrove) [reply docs/source/contributor-guide/plugin_overview.md]: @sunchao Thanks. I have updated this. Let me know what you think.

Comment (sunchao) [reply docs/source/contributor-guide/plugin_overview.md]: Looks good. Thanks

File: docs/source/contributor-guide/plugin_overview.md (hunk 1 above)
Comment (parthchandra) [reply docs/source/contributor-guide/plugin_overview.md]: Sorry, just seeing this today. The `explain` information will show up in the UI or `EXPLAIN` output only from Spark 4.0.0 onwards as the `ExtendedExplainGenerator` trait was only added in Spark 4.0. Internally, we can always call `ExtendedExplainInfo.generateExtendedInfo(plan)`

**Pull request merged**